protocol/src/main/java/org/fido/iot/protocol/CryptoService.java

```diff
@@ -1393,94 +1385,46 @@ protected Composite encryptThenMac(byte[] secret, byte[] ciphered, byte[] iv, St
-    byte[] sek = new byte[Const.BIT_LEN_256 / Byte.SIZE]; // 32 bytes
-    buffer.get(sek);
-
-    //HMAC-SHA-256[0,(byte)2||"FIDO-KDF"||(byte)0||"AutomaticOnboard-hmac"||ShSe]
-    buffer = ByteBuffer
-        .allocate(1 + Const.KDF_STRING.length + 1 + Const.PROV_HMAC.length + shSe.length);
-
-    buffer.put((byte) 2);
-    buffer.put(Const.KDF_STRING);
-    buffer.put((byte) 0);
-    buffer.put(Const.PROV_HMAC);
-    buffer.put(shSe);
-    buffer.flip();
-    hash = hash(Const.HMAC_SHA_384, Const.HMAC_ZERO, Composite.unwrap(buffer));
-    buffer = ByteBuffer.wrap(hash.getAsBytes(Const.HASH));
-    byte[] svk1 = new byte[Const.BIT_LEN_384 / Byte.SIZE]; //48 bytes
-    buffer.get(svk1);
-
-    //HMAC-SHA-256[0,(byte)3||"FIDO-KDF"||(byte)0||"AutomaticOnboard-hmac"||ShSe]
-    buffer = ByteBuffer
-        .allocate(1 + Const.KDF_STRING.length + 1 + Const.PROV_HMAC.length + shSe.length);
-    buffer.put((byte) 3);
-    buffer.put(Const.KDF_STRING);
-    buffer.put((byte) 0);
-    buffer.put(Const.PROV_HMAC);
-    buffer.put(shSe);
-    buffer.flip();
-    hash = hash(Const.HMAC_SHA_384, Const.HMAC_ZERO, Composite.unwrap(buffer));
-    buffer = ByteBuffer.wrap(hash.getAsBytes(Const.HASH));
-    byte[] svk2 = new byte[(Const.BIT_LEN_256 / Byte.SIZE) / 2];//16 bytes
-    buffer.get(svk2);
-
-    buffer = ByteBuffer.allocate(svk1.length + svk2.length);
-    buffer.put(svk1);
-    buffer.put(svk2);
-    buffer.flip();
+  // Key Derivation Function (KDF).
+  //
+  // See NIST SP 800-108, FDO spec section 3.6.4
+  // Where possible, variable names are chosen to match those documents.
+  protected byte[] kdf(
+      int size,      // the number of bytes to derive (L)
+      String prfId,  // the JCE ID of the PRF to use
+      KeyExchangeResult kxResult) // the sharedSecret and contextRandom
+      throws
+      InvalidKeyException,
+      IOException,
+      NoSuchAlgorithmException {
+
+    Mac prf = Mac.getInstance(prfId);
+    prf.init(new SecretKeySpec(kxResult.shSe, prfId));
+
+    final int h = prf.getMacLength(); // (h) the length (in bits) of the output of the PRF
+    final int l = size;  // (L) the length (in bits) of the derived keying material
+    // (n) the number of iterations of the PRF needed to generate L bits of
+    // keying material.  We count in bytes, but the result is the same.
+    final int n = Double.valueOf(Math.ceil((double)l / (double)h)).intValue();
+
+    ByteArrayOutputStream result = new ByteArrayOutputStream();
+
+    for (int i = 0; n > i; i++) {
+
+      prf.reset();
+
+      // write K(i) to the prf...
+      prf.update((byte)i); // [i]2
+      prf.update("FIDO-KDF".getBytes(StandardCharsets.UTF_8)); // Label
+      prf.update((byte)0); // 0x00, separator
+      prf.update("AutomaticOnboardTunnel".getBytes(StandardCharsets.UTF_8)); // Context (part 1)
+      prf.update(kxResult.contextRand);                                      // Context (part 2)
+      prf.update((byte)((l >> 8) & 0xff)); // [L]2, upper byte
```

Comment:
Isn't it same as `prf.update((byte)0)` ?

Reply:
No, because if more than 255 bits of key material is needed, there will be bits in the upper byte.

Reply:
My bad .. was reading `l` as 1.